File: app/src/main/java/de/gymnasium_lappersdorf/gymlapapp/HausaufgabenPlaner/HausaufgabenLokalFragment.java
package de.gymnasium_lappersdorf.gymlapapp.HausaufgabenPlaner;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

import de.gymnasium_lappersdorf.gymlapapp.R;

import static android.app.Activity.RESULT_OK;

/**
 * 07.12.2018 | created by Lukas S
 */
public class HausaufgabenLokalFragment extends HausaufgabenTabFragment {

    private static final int REQUEST_ID = 1;

    private ItemTouchHelper itemTouchHelper;

    private Snackbar snackbarRevert;

    //Recycler swipe removing
    private Hausaufgabe lastItem = null;
    private int lastItemPosition = -1;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        setRecyclerView((RecyclerView) getV().findViewById(R.id.homework_rv));

        setHomeworkRvAdapter(new HomeworkRvAdapter(new Hausaufgabe[0], getActivity(), new HomeworkRvAdapter.DatasetChangeListener() {
            @Override
            public void onNotificationIdChanged(Hausaufgabe h) {
                int index = getHomeworks().indexOf(h);
                getHomeworks().get(index).setNotificationId(h.getNotificationId());
                getDbh().updateHomework(getHomeworks().get(index));
            }
        }));
        getRecyclerView().setAdapter(getHomeworkRvAdapter());

        setLinearLayoutManager(new LinearLayoutManager(getActivity()));
        getRecyclerView().setLayoutManager(getLinearLayoutManager());

        setCountLabel((TextView) getV().findViewById(R.id.homework_count_label));


        setFab((FloatingActionButton) getV().findViewById(R.id.add_homework_fab));
        getFab().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(getActivity(), AddHomeworkActivity.class);
                i.putExtra("STUFE", getStufe());
                i.putExtra("KLASSE", getKlasse() == getKlassenArray()[0] ? "a" : getKlasse());
                startActivityForResult(i, REQUEST_ID);
            }
        });
        setFabContainer((CoordinatorLayout) getV().findViewById(R.id.fab_container));

        snackbarRevert = Snackbar.make(getFabContainer(), "Hausaufgabe erledigt", Snackbar.LENGTH_LONG);
        snackbarRevert.setAction("Rückgängig", new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (lastItem != null) {
                    int pos = getHomeworks().indexOf(lastItem);
                    getHomeworks().get(pos).setDone(false);
                    getHomeworkRvAdapter().restoreItem(getHomeworks().get(pos), lastItemPosition);
                    updateLabel();
                    getRecyclerView().smoothScrollToPosition(lastItemPosition);
                    lastItem = null;
                    //Updating database
                    getDbh().updateHomework(getHomeworks().get(pos));
                }
            }
        });

        //Swipe card to remove
        ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int swipeDir) {
                int pos = getHomeworks().indexOf(getHomeworkRvAdapter().getDataset()[viewHolder.getAdapterPosition()]);
                getHomeworks().get(pos).setDone(true);
                lastItem = getHomeworks().get(pos);
                lastItemPosition = viewHolder.getAdapterPosition();
                getHomeworkRvAdapter().removeItem(viewHolder.getAdapterPosition());
                updateLabel();
                snackbarRevert.show();
                //Updating database
                getDbh().updateHomework(getHomeworks().get(pos));
            }
        };

        itemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);
        itemTouchHelper.attachToRecyclerView(getRecyclerView());

        setStufe(5);
        setKlasse(getKlassenArray()[0]);

        updateDataset();
        updateLabel();
    }

    //Adding new homework to "homeworks"
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {

        if (requestCode == REQUEST_ID) {

            if (resultCode == RESULT_OK) {
                long id = data.getExtras().getLong("HW_ID");

                Hausaufgabe temp = getDbh().getHomework(id);

                getHomeworks().add(temp);

                filter(temp.getStufe(), temp.getKurs());
            }

        }
    }

    @Override
    public void updateDataset() {
        setDbh(new HausaufgabenDatabaseHandler(getActivity()));
        setHomeworks(new ArrayList<Hausaufgabe>());

        if (getDbh().getHomeworkCount() > 0) {
            //load content of db
            Hausaufgabe[] hw_db = getDbh().getAllHomeworks();

            Collections.addAll(getHomeworks(), hw_db);

            //Removing all local homeworks that are marked as done and all online homeworks
            Iterator it = getHomeworks().iterator();
            Hausaufgabe temp;
            while (it.hasNext()) {
                temp = (Hausaufgabe) it.next();
                if (temp.isDone() && !temp.isFromInternet()) {
                    it.remove();
                    getDbh().deleteHomework(temp);
                } else if (temp.isFromInternet()) {
                    it.remove();
                }
            }
        } else {
            setHomeworks(new ArrayList<Hausaufgabe>()); //TODO: Remove
        }

        filter(getStufe(), getKlasse());
    }

}

File: app/src/main/java/de/gymnasium_lappersdorf/gymlapapp/HausaufgabenPlaner/HausaufgabenLokalFragment.kt
package de.gymnasium_lappersdorf.gymlapapp.HausaufgabenPlaner

import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import de.gymnasium_lappersdorf.gymlapapp.R
import java.util.*

/**
 * 07.12.2018 | created by Lukas S
 */
class HausaufgabenLokalFragment : Fragment() {

    private val REQUEST_ID = 1
    private val EDIT_ID = 2

    private var homeworkList: ArrayList<Hausaufgabe>? = null

    private lateinit var v: View
    private var countLabel: TextView? = null
    private var recyclerView: RecyclerView? = null
    private var linearLayoutManager: LinearLayoutManager? = null
    private var homeworkRvAdapter: HomeworkRvAdapter? = null

    private var fab: FloatingActionButton? = null
    private var fabContainer: CoordinatorLayout? = null

    private var itemTouchHelper: ItemTouchHelper? = null

    //Database
    private var dbh: HausaufgabenDatabaseHandler? = null

    //Recycler swipe removing
    private var lastItem: Hausaufgabe? = null
    private var lastItemPosition = -1
    private var snackbarRevert: Snackbar? = null

    fun getRequestId() = REQUEST_ID
    fun getEditId() = EDIT_ID

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        v = inflater.inflate(R.layout.fragment_homework_tab_local, container, false)
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)


        recyclerView = v.findViewById<View>(R.id.homework_rv) as RecyclerView

        homeworkRvAdapter = HomeworkRvAdapter(arrayOfNulls(0), activity,this, HomeworkRvAdapter.DatasetChangeListener { h ->
            val index = homeworkList!!.indexOf(h)
            homeworkList!![index].notificationId = h.notificationId
            dbh!!.updateHomework(homeworkList!![index])
        })
        recyclerView!!.adapter = homeworkRvAdapter

        linearLayoutManager = LinearLayoutManager(activity)
        recyclerView!!.layoutManager = linearLayoutManager

        countLabel = v.findViewById<View>(R.id.homework_count_label) as TextView


        fab = v.findViewById<View>(R.id.add_homework_fab) as FloatingActionButton
        fab!!.setOnClickListener {
            val i = Intent(activity, AddHomeworkActivity::class.java)
            startActivityForResult(i, REQUEST_ID)
        }
        fabContainer = v.findViewById<View>(R.id.fab_container) as CoordinatorLayout

        snackbarRevert = Snackbar.make(fabContainer!!, "Hausaufgabe erledigt", Snackbar.LENGTH_LONG)
        snackbarRevert!!.setAction("Rückgängig") {
            if (lastItem != null) {
                val pos = homeworkList!!.indexOf(lastItem!!)
                homeworkList!![pos].isDone = false
                homeworkRvAdapter!!.restoreItem(homeworkList!![pos], lastItemPosition)
                updateLabel()
                recyclerView!!.smoothScrollToPosition(lastItemPosition)
                lastItem = null
                //Updating database
                dbh!!.updateHomework(homeworkList!![pos])
            }
        }

        //Swipe card to remove
        val simpleItemTouchCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
                val pos = homeworkList!!.indexOf(homeworkRvAdapter!!.dataset[viewHolder.adapterPosition])
                homeworkList!![pos].isDone = true
                lastItem = homeworkList!![pos]
                lastItemPosition = viewHolder.adapterPosition
                homeworkRvAdapter!!.removeItem(viewHolder.adapterPosition)
                updateLabel()
                snackbarRevert!!.show()
                //Updating database
                dbh!!.updateHomework(homeworkList!![pos])
            }
        }

        itemTouchHelper = ItemTouchHelper(simpleItemTouchCallback)
        itemTouchHelper!!.attachToRecyclerView(recyclerView)

        updateDataset()
        updateLabel()
    }

    private fun updateLabel() {
        val count = homeworkRvAdapter!!.itemCount
        countLabel!!.text = "${(if (count == 0) "Kein" else homeworkRvAdapter!!.itemCount)}${if (homeworkRvAdapter!!.itemCount > 1) " Einträge" else " Eintrag"} vorhanden."
    }

    //Adding new homework to "homeworkList"
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        if (resultCode == RESULT_OK) {

            if (requestCode == REQUEST_ID) {
                val id = data!!.extras!!.getLong("HW_ID")

                val temp = dbh!!.getHomework(id)

                homeworkList!!.add(temp)

                homeworkRvAdapter!!.dataset = homeworkList!!.toTypedArray()
                homeworkRvAdapter!!.notifyDataSetChanged()
                updateLabel()
            } else if (requestCode == EDIT_ID) {
                val id = data!!.extras!!.getLong("HW_ID")

                updateDataset()
            }

        }
    }

    private fun updateDataset() {
        dbh = HausaufgabenDatabaseHandler(activity)
        homeworkList = ArrayList()

        if (dbh!!.homeworkCount > 0) {
            //load content of db
            val hw_db = dbh!!.completeHomework

            Collections.addAll(homeworkList, *hw_db)

            //Removing all local homeworkList that are marked as done and all online homeworkList
            val it = homeworkList!!.iterator()
            var temp: Hausaufgabe
            while (it.hasNext()) {
                temp = it.next()
                if ((temp.isDone && !temp.isFromInternet) || temp.isOutdated) {
                    it.remove()
                    dbh!!.deleteHomework(temp)
                } else if (temp.isFromInternet) {
                    it.remove()
                }
            }
        } else {
            homeworkList = ArrayList()
        }

        homeworkRvAdapter!!.dataset = homeworkList!!.toTypedArray()
        homeworkRvAdapter!!.notifyDataSetChanged()
        updateLabel()
    }

}
